test/apps/iodemo/io_demo.cc

```diff
@@ -265,15 +323,26 @@ class DemoServer : public P2pDemoCommon {
             LOG << "Invalid opcode: " << hdr->op;
         }
     }
+public:
+    static MemoryPool<IoWriteResponseCallback> wCallbackPool;    
 };
+MemoryPool<DemoServer::IoWriteResponseCallback> DemoServer::wCallbackPool;
 
 
 class DemoClient : public P2pDemoCommon {
 public:
     class IoReadResponseCallback : public UcxCallback {
     public:
-        IoReadResponseCallback(long *counter, size_t iomsg_size) :
-            _counter(0), _io_counter(counter), _buffer(malloc(iomsg_size)) {
+        IoReadResponseCallback() :
+            _counter(0), _io_counter(0), _buffer(NULL) {
+        }
+        
+        void init(long *counter, size_t iomsg_size) {
+            _counter    = 0;
+            _io_counter = counter;
+            if (!_buffer){
+                _buffer = malloc(iomsg_size); 
```

Comment:
We need r->buffer() in do_io_read() function.

Reply:
can we allocate the buffer in constructor?

Reply:
> can we allocate the buffer in constructor?

Maybe not. If we allocate the buffer in constructor, then we should put the size of the buffer as parameter of constructor. The constructor function is excuted in MemoryPool::get(). We don't want to deal with the buffer size in MemoryPool class since this is not a concept of MemoryPool class. MemoryPool just allocate the items, but don't care the allocation of buffers(other resource) of items.